src/bin/cargo-miri.rs

```diff
@@ -34,6 +48,145 @@ fn show_version() {
         env!("CARGO_PKG_VERSION"), env!("VERGEN_SHA_SHORT"), env!("VERGEN_COMMIT_DATE"));
 }
 
+fn show_error(msg: String) -> ! {
+    eprintln!("fatal error: {}", msg);
+    std::process::exit(1)
+}
+
+fn list_targets(mut args: impl Iterator<Item=String>) -> impl Iterator<Item=cargo_metadata::Target> {
+    // We need to get the manifest, and then the metadata, to enumerate targets.
+    let manifest_path_arg = args.find(|val| {
+        val.starts_with("--manifest-path=")
+    });
+
+    let mut metadata = if let Ok(metadata) = cargo_metadata::metadata(
+        manifest_path_arg.as_ref().map(AsRef::as_ref),
+    )
+    {
+        metadata
+    } else {
+        let _ = std::io::stderr().write_fmt(format_args!(
+            "error: Could not obtain cargo metadata."
+        ));
+        std::process::exit(101);
+    };
+
+    let manifest_path = manifest_path_arg.map(|arg| {
+        PathBuf::from(Path::new(&arg["--manifest-path=".len()..]))
+    });
+
+    let current_dir = std::env::current_dir();
+
+    let package_index = metadata
+        .packages
+        .iter()
+        .position(|package| {
+            let package_manifest_path = Path::new(&package.manifest_path);
+            if let Some(ref manifest_path) = manifest_path {
+                package_manifest_path == manifest_path
+            } else {
+                let current_dir = current_dir.as_ref().expect(
+                    "could not read current directory",
+                );
+                let package_manifest_directory = package_manifest_path.parent().expect(
+                    "could not find parent directory of package manifest",
+                );
+                package_manifest_directory == current_dir
+            }
+        })
+        .expect("could not find matching package");
+    let package = metadata.packages.remove(package_index);
+
+    // Finally we got the list of targets to build
+    package.targets.into_iter()
+}
+
+fn ask(question: &str) {
+    let mut buf = String::new();
+    print!("{} [Y/n] ", question);
+    io::stdout().flush().unwrap();
+    io::stdin().read_line(&mut buf).unwrap();
+    let answer = match buf.trim().to_lowercase().as_ref() {
+        "" | "y" | "yes" => true,
+        "n" | "no" => false,
+        a => show_error(format!("I do not understand `{}`", a))
+    };
+    if !answer {
+        show_error(format!("Aborting as per your request"))
```

Comment:
could you move this to the match case returning false?

Reply:
Hm, I found it nicer to first convert-to-bool and then act on the bool. But I do not particularly care.

Reply:
I also think that it should be in the match.